Tasker.cpp

```diff
@@ -0,0 +1,258 @@
+/**
+ * @file SimpleTasker.cpp
+ * @author Jan Wielgus & Aleksy Walczak
+ * @brief 
+ * @date 2020-07-31
+ * 
+ */
+
+#include "Tasker.h"
+
+
+const uint32_t Tasker::MinTaskInterval_us = 52;
+
+Tasker::Tasker(uint8_t maxTasksAmount)
+    : MaxAmtOfTasks(maxTasksAmount)
+{
+    if (MaxAmtOfTasks > 0)
+        tasks = new Task[MaxAmtOfTasks];
+}
+
+
+Tasker::~Tasker()
+{
+    if (MaxAmtOfTasks > 0)
+        delete[] tasks;
+}
+
+
+bool Tasker::addTask_Hz(IExecutable* task, float frequency_Hz)
+{
+    if (amountOfTasks >= MaxAmtOfTasks || task == nullptr || frequency_Hz <= 0)
+        return false;
+
+    Task newTask;
+    newTask.executable = task;
+
+    uint32_t newInterval_us = 1000000.0 / frequency_Hz + 0.5f;
+    newTask.interval_us = newInterval_us < MinTaskInterval_us ? MinTaskInterval_us : newInterval_us;
+
+    newTask.nextExecutionTime_us = currentTime;
+
+    tasks[amountOfTasks++] = newTask;
+
+    calculateNextTask();
+    return true;
+}
+
+
+bool Tasker::addTask_us(IExecutable* task, uint32_t interval_us)
+{
+    if (amountOfTasks >= MaxAmtOfTasks || task == nullptr)
+        return false;
+
+    Task newTask;
+    newTask.executable = task;
+    newTask.interval_us = interval_us < MinTaskInterval_us ? MinTaskInterval_us : interval_us;
+    newTask.nextExecutionTime_us = currentTime;
+
+    tasks[amountOfTasks++] = newTask;
+
+    calculateNextTask();
+    return true;
+}
+
+
+bool Tasker::removeTask(IExecutable* task)
+{
+    bool isTaskFound = false;
+    for (uint8_t i = 0; i < amountOfTasks-1; ++i)
+    {
+        if (tasks[i].executable == task)
+            isTaskFound = true;
+        
+        if (isTaskFound)
+            tasks[i] = tasks[i+1];
+    }
+
+    if (tasks[amountOfTasks-1].executable == task)
+        isTaskFound = true;
+
+    if (isTaskFound)
+        amountOfTasks--;
+    
+    calculateNextTask();
+
+    return isTaskFound;
+}
+
+
+void Tasker::calculateNextTask()
+{
+    if (amountOfTasks == 0)
+    {
+        nextTask = nullptr;
+        return;
+    }
+    
+    nextTask = tasks;
+    for (uint8_t i = 1; i < amountOfTasks; ++i)
+    {
+        if (tasks[i].nextExecutionTime_us < nextTask->nextExecutionTime_us)
+            nextTask = &tasks[i];
+    }
+}
+
+
+bool Tasker::setTaskFrequency(IExecutable* task, float frequency_Hz)
+{
+    for (uint8_t i = 0; i < amountOfTasks-1; ++i)
```

Comment:
without -1 (but irrelevant if method body will be replaced with setTaskInterval_us() method call)